apps/src/gamelab/GameLab.js

```diff
@@ -0,0 +1,574 @@
+        this.currentCmdQueue = handler.cmdQueue;
+        try {
+          handler.func(this.studioApp_, this.api, this.Globals);
+        } catch (e) {
+          // Do nothing
+        }
+        this.currentCmdQueue = null;
+      }
+    } else {
+      // TODO (cpirich): support events with parameters
+      if (handler.name === name) {
+        this.JSInterpreter.queueEvent(handler.func, extraArgs);
+      }
+    }
+  }, this));
+};
+
+//
+// Execute an entire command queue (specified with the name parameter)
+//
+
+GameLab.prototype.executeQueue = function (name, oneOnly) {
+  this.eventHandlers.forEach(_.bind(function (handler) {
+    if (handler.name === name && handler.cmdQueue.length) {
+      for (var cmd = handler.cmdQueue[0]; cmd; cmd = handler.cmdQueue[0]) {
+        if (this.callCmd(cmd)) {
+          // Command executed immediately, remove from queue and continue
+          handler.cmdQueue.shift();
+        } else {
+          break;
+        }
+      }
+    }
+  }, this));
+};
+
+GameLab.prototype.evalCode = function(code) {
+  try {
+    codegen.evalWith(code, {
+      GameLab: this.api
+    });
+  } catch (e) {
+    // Infinity is thrown if we detect an infinite loop. In that case we'll
+    // stop further execution, animate what occured before the infinite loop,
+    // and analyze success/failure based on what was drawn.
+    // Otherwise, abnormal termination is a user error.
+    if (e !== Infinity) {
+      // call window.onerror so that we get new relic collection.  prepend with
+      // UserCode so that it's clear this is in eval'ed code.
+      if (window.onerror) {
+        window.onerror("UserCode:" + e.message, document.URL, 0);
+      }
+      window.alert(e);
+    }
+  }
+};
+
+/**
+ * Set up this.code, this.interpreter, etc. to run code for editCode levels
+ */
+
+/*
+GameLab.prototype.generateTurtleCodeFromJS_ = function () {
+  this.code = dropletUtils.generateCodeAliases(dropletConfig, 'Turtle');
+  this.userCodeStartOffset = this.code.length;
+  this.code += this.studioApp_.editor.getValue();
+  this.userCodeLength = this.code.length - this.userCodeStartOffset;
+
+  var session = this.studioApp_.editor.aceEditor.getSession();
+  this.cumulativeLength = codegen.aceCalculateCumulativeLength(session);
+
+  var initFunc = _.bind(function(interpreter, scope) {
+    codegen.initJSInterpreter(interpreter, null, null, scope, {
+      Turtle: this.api
+    });
+  }, this);
+  this.interpreter = new window.Interpreter(this.code, initFunc);
+};
+*/
+
+/**
+ * Execute the user's code.  Heaven help us...
+ */
+GameLab.prototype.execute = function() {
+  // Reset all state.
+  this.studioApp_.reset();
+
+  if (this.studioApp_.isUsingBlockly() &&
+      (this.studioApp_.hasExtraTopBlocks() ||
+        this.studioApp_.hasDuplicateVariablesInForLoops())) {
+    // immediately check answer, which will fail and report top level blocks
+    this.checkAnswer();
+    return;
+  }
+
+  if (this.level.editCode) {
```

Comment:
I wonder if these sort of if-statements should be considered a code smell in a new app type.  If we want Gamelab to be Blockly/Droplet agnostic as much as possible, this is probably the right time to figure out how to hide them behind a common interface so that we can essentially "inject" one or the other into Gamelab and have everything work.

Thoughts?